compiler/rustc_llvm/llvm-wrapper/CoverageMappingWrapper.cpp

```diff
@@ -58,15 +62,72 @@ fromRust(LLVMRustCounterMappingRegionKind Kind) {
     return coverage::CounterMappingRegion::GapRegion;
   case LLVMRustCounterMappingRegionKind::BranchRegion:
     return coverage::CounterMappingRegion::BranchRegion;
+  case LLVMRustCounterMappingRegionKind::MCDCDecisionRegion:
+    return coverage::CounterMappingRegion::MCDCDecisionRegion;
+  case LLVMRustCounterMappingRegionKind::MCDCBranchRegion:
+    return coverage::CounterMappingRegion::MCDCBranchRegion;
   }
   report_fatal_error("Bad LLVMRustCounterMappingRegionKind!");
 }
 
+enum class LLVMRustMCDCParametersTag : uint8_t {
+  None = 0,
+  Decision = 1,
+  Branch = 2,
+};
+
+struct LLVMRustMCDCDecisionParameters {
+  uint32_t BitmapIdx;
+  uint16_t NumConditions;
+};
+
+struct LLVMRustMCDCBranchParameters {
+  int16_t ConditionID;
+  int16_t ConditionIDs[2];
+};
+
+union LLVMRustMCDCParametersPayload {
+  LLVMRustMCDCDecisionParameters DecisionParameters;
+  LLVMRustMCDCBranchParameters BranchParameters;
+};
+
+struct LLVMRustMCDCParameters {
+  LLVMRustMCDCParametersTag Tag;
+  LLVMRustMCDCParametersPayload Payload;
+};
```

Comment:
We should design our FFI structures in a way that makes the C++ code as simple and obviously-correct as possible, even if it's a bit inefficient and requires some extra work on the Rust side.

For example, I think we should try to avoid using `union` at all.

I'm not 100% sure about this, but maybe we could just have one `LLVMRustMCDCParameters` struct that has all the fields for both variants, with the unused fields being set to zero and not read.

Or if that feels too large, we could have one struct with enough fields to handle both cases, and use the fields for different things depending on the tag.

What do you think?

Reply:
Looking at LLVM's `struct MCDCParameters`, even they just have one big struct with all the fields. So we should avoid doing anything more complicated than that.

Reply:
Actually, I think we can just reuse a lot of the FFI stuff from #123358, since that PR did a cleaner job in this area.

Reply:
See https://github.com/rust-lang/rust/pull/123358/files#r1552756599 for some notes.

Reply:
The llvm MCDCParameter has changed from 18 (which we see in rust backend) to 19 (to a `std::variant`). The rust side structure mostly looks like 19, including the namespace style, see [`MCDCTypes.h`](https://github.com/llvm/llvm-project/blob/843cc474faefad1d639f4c44c1cf3ad7dbda76c8/llvm/include/llvm/ProfileData/Coverage/MCDCTypes.h#L51).

And also we do not familiar with the convention of llvm wrapper for rustc. Just check it is valid according to [unsafe guideline](https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html#explicit-repr-annotation-with-c-compatibility).

But I think it's very reasonable to avoid unions indeed. We may make the struct more plain such as 
```rust
#[repr(C)]
struct LLVMRustMCDCParameters {
  tag: LLVMRustMCDCParametersTag,
  decision_parameter: LLVMRustMCDCDecisionParameter,
  condition_parameter: LLVMRustMCDCBranchParameter
}
```
Is it more appropriate？

Reply:
What's more I have learnt that rustc should support llvm which is released version-1. So there is a build check for llvm 17 now. Also once rustc upgraded its llvm to 19 we also should support llvm-18. Here we should add open gates to deal with the different llvm representations.

Reply:
Ah, I see. I hadn't looked at the LLVM 19 version.

In that case, I think it's OK to use FFI types that are closer to the 19 version, and then have C++ code that adapts them to the 18 version. (But we should have comments that explain why this is happening, and link to the LLVM 19 data structures.)

For the struct, I think your suggestion looks like a good idea. The main priority is to make it very easy to review/audit, and I think it's OK to “waste” a little bit of memory by having both structures nested inside `LLVMRustMCDCParameters`, even though only 0/1 will be used at a time.

Reply:
I think I'll ask on Zulip to see what other people think about the union vs struct question.

Reply:
https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/Passing.20union-like.20data.20.28MC.2FDC.20coverage.29.20to.20LLVM.20across.20FFI

Reply:
I have changed the representation so that it presents in a more c-like way.